File: system-design/oop-patterns/06-singleton-pattern/final/ts/main.ts
// Singleton Pattern - final
//
// The class-based getInstance() from the starter file works fine in TS,
// there's no race to fix. But TypeScript/JavaScript has an even simpler
// idiomatic answer: a module is only ever evaluated once and then cached,
// so exporting a single instance directly IS the singleton, no class,
// no getInstance() method, no private constructor needed at all.

class ConnectionPoolImpl {
  connect(): string {
    return 'connected';
  }
}

export const connectionPool = new ConnectionPoolImpl();

// Every import of this module, anywhere in the app, receives this exact
// same object, because the module itself only ever runs once.
console.log(connectionPool.connect());

File: system-design/oop-patterns/06-singleton-pattern/final/cpp/main.cpp
// Singleton Pattern - final
//
// ConnectionPool now uses a function-local static, which C++11 guarantees
// is initialized exactly once, even under concurrent calls, with no manual
// lock needed. Every thread gets back the exact same address, every run.

#include <iostream>
#include <thread>
#include <vector>

class ConnectionPool {
public:
    static ConnectionPool& getInstance() {
        static ConnectionPool instance;
        return instance;
    }

private:
    ConnectionPool() = default;
};

int main() {
    const int threadCount = 8;
    std::vector<std::thread> threads;
    std::vector<ConnectionPool*> results(threadCount);

    for (int i = 0; i < threadCount; ++i) {
        threads.emplace_back([&results, i]() {
            results[i] = &ConnectionPool::getInstance();
        });
    }

    for (auto& t : threads) t.join();

    for (ConnectionPool* pool : results) {
        std::cout << pool << "\n";
    }

    return 0;
}
